onboarding/mlsys/1-gradients.md

```diff
@@ -0,0 +1,494 @@
+|     (1, 0)    |              |
+|     (2, 1)    |              |
+|     (3, 1)    |              |
+|    (4, 6.25)  |              |
+
+Then, with `m=2`, fill the following table of `batch_size=2` batch gradients:
+
+| `(x0, ytrue0)` | `(x1, ytrue1)` | `grad_multi` |
+| -------------- | -------------- | ------------ |
+|     (1, 0)     |     (2, 1)     |              |
+|     (3, 1)     |    (4, 6.25)   |              |
+
+Do you see that by averaging gradients from multiple points, we end up with
+smoother overall gradients?
+
+### Q12
+
+Write a new training loop that takes batches instead of individual points.
+
+The only difference is that instead of using the gradient from one point per
+iteration, you will be using the gradient of a batch of points for every
+iteration.  You should assume that the dataset is already broken up into
+batches, and your train function does not need to handle that.
+
+Here are some test cases:
+
+```python
+dataset = [
+    [(1,0),(2,1)],
+    [(3,1),(4,6.25)],
+]
+mfinal = train_loop_batches(0, lr=0.001, dataset=dataset, iterations=2)
+assert round(mfinal, 5) == 0.02995, mfinal
+mfinal = train_loop_batches(0, lr=0.001, dataset=dataset, iterations=20)
+assert round(mfinal, 5) == 0.26228, mfinal
+```
+
+## Averaging Gradients Together Means Faster Training
+
+Here is the central idea of this lesson: **when you average gradients from
+multiple data points, you get less noisy gradients, which lets you train
+faster**.  Less noisy gradients means you have higher confidence that the step
+you take after calculating the gradient is in the correct direction.  Higher
+confidence in the gradient direction means you can take bigger steps (by
+increasing your learning rate).  A higher learning rate means your model
+converges faster.
+
+This is really important.  This is foundational to understanding and supporting
+our product offering.
+
+The general rule of thumb for Stochastic Gradient Descent (basically the
+algorithm you just implemented) is that you can scale up your learning rate
+linearly as you increase your batch size.
+
+There are hardware upper limits to how far you can push that rule of thumb, of
+course.  You can only train efficiently on what fits in RAM, and the larger the
+batch size, the more memory space is required, both for the data itself and for
+various intermediate values used in training.
+
+## Data-Parallel Distributed Training
+
+Suppose you have a model that takes days or weeks to train on a single GPU, and
+you'd like to throw more hardware at it and get it to train faster.
+
+Any time you train a single model on multiple accelerators, that's called
+Distributed Training (or "dtrain", in internal company slang).
+
+There are a few kinds of dtrain, but we'll focus on data-parallel dtrain,
+because it's more common, and also it's a straightforward application of
+averaging gradients together, like we've been doing.
+
+The basic strategy of data-parallel d-train is you have one worker per
```

Comment:
I didn't see any other usages of "d-train" but to keep it consistent, use "dtrain".